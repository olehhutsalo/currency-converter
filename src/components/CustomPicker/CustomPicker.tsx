import React, { useState } from 'react'

import {
  View,
  Picker,
  Modal,
  Button,
  TouchableOpacity,
  Text,
  Platform,
  StyleSheet,
  ViewStyle,
  TextStyle,
} from 'react-native'
import { AntDesign } from '@expo/vector-icons'

export type PickerItemType = { label: string; value: any }
type CustomPickerPropTypes = {
  items: PickerItemType[]
  selectedValue: any
  onChange: (value: any) => any
  styles?: { input?: ViewStyle; label?: TextStyle }
}

const isIos = Platform.OS === 'ios'

const defaultStyles = StyleSheet.create({
  picker: {
    minWidth: 100,
    padding: 16,
    backgroundColor: '#fff',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  label: { marginRight: 4 },
  modalContainer: {
    flex: 1,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    justifyContent: 'center',
  },
  modalContentContainer: {
    backgroundColor: '#fff',
    borderRadius: 8,
    overflow: 'hidden',
  },
})

export default ({
  items,
  selectedValue,
  onChange,
  styles = {},
}: CustomPickerPropTypes) => {
  const [modalVisible, setModalVisible] = useState(false)

  const selectedItem = items.find(
    (item: PickerItemType) => item.value === selectedValue,
  )

  const selectedLabel = selectedItem ? selectedItem.label : ''

  if (isIos) {
    return (
      <React.Fragment>
        <TouchableOpacity onPress={() => setModalVisible(true)}>
          <View style={[defaultStyles.picker, styles.input]}>
            <Text style={[defaultStyles.label, styles.label]}>
              {selectedLabel}
            </Text>
            <AntDesign name='down' size={18} />
          </View>
        </TouchableOpacity>
        <Modal visible={modalVisible} onRequestClose={() => {}} transparent>
          <View style={defaultStyles.modalContainer}>
            <View style={defaultStyles.modalContentContainer}>
              <Picker
                selectedValue={selectedValue}
                onValueChange={(itemValue) => {
                  onChange(itemValue)
                }}
              >
                {items.map(({ label, value }: PickerItemType, i: number) => (
                  <Picker.Item key={`item${i}`} label={label} value={value} />
                ))}
              </Picker>
              <Button title='Done' onPress={() => setModalVisible(false)} />
            </View>
          </View>
        </Modal>
      </React.Fragment>
    )
  }
  return (
    <View>
      <View>
        <Picker
          selectedValue={selectedValue}
          onValueChange={onChange}
          mode='dropdown'
        >
          {items.map(({ label, value }: PickerItemType, i: number) => (
            <Picker.Item key={`item${i}`} label={label} value={value} />
          ))}
        </Picker>
      </View>
    </View>
  )
}
