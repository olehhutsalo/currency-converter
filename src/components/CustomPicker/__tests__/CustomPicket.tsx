import React from 'react'
import CustomPicker from '../CustomPicker'
import { shallow } from 'enzyme'

const items = [
  { label: 'Foo', value: 'foo' },
  { label: 'Bar', value: 'bar' },
]

describe('components -> <CustomPicker />', () => {
  it('renders with selected value on iOS', () => {
    const component = shallow(
      <CustomPicker items={items} selectedValue={'bar'} onChange={() => {}} />,
    )
    expect(component).toMatchSnapshot()
  })

  it('renders without selected value on iOS', () => {
    const component = shallow(
      <CustomPicker items={items} selectedValue={null} onChange={() => {}} />,
    )
    expect(component).toMatchSnapshot()
  })
})
