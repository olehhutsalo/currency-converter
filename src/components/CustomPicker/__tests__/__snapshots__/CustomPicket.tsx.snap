// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components -> <CustomPicker /> renders with selected value on iOS 1`] = `
<Fragment>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "minWidth": 100,
            "padding": 16,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "marginRight": 4,
            },
            undefined,
          ]
        }
      >
        Bar
      </Text>
      <Icon
        allowFontScaling={false}
        name="down"
        size={18}
      />
    </View>
  </TouchableOpacity>
  <Component
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.5)",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#fff",
            "borderRadius": 8,
            "overflow": "hidden",
          }
        }
      >
        <Picker
          mode="dialog"
          onValueChange={[Function]}
          selectedValue="bar"
        >
          <PickerItem
            key="item0"
            label="Foo"
            value="foo"
          />
          <PickerItem
            key="item1"
            label="Bar"
            value="bar"
          />
        </Picker>
        <Button
          onPress={[Function]}
          title="Done"
        />
      </View>
    </View>
  </Component>
</Fragment>
`;

exports[`components -> <CustomPicker /> renders without selected value on iOS 1`] = `
<Fragment>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "minWidth": 100,
            "padding": 16,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "marginRight": 4,
            },
            undefined,
          ]
        }
      />
      <Icon
        allowFontScaling={false}
        name="down"
        size={18}
      />
    </View>
  </TouchableOpacity>
  <Component
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.5)",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#fff",
            "borderRadius": 8,
            "overflow": "hidden",
          }
        }
      >
        <Picker
          mode="dialog"
          onValueChange={[Function]}
          selectedValue={null}
        >
          <PickerItem
            key="item0"
            label="Foo"
            value="foo"
          />
          <PickerItem
            key="item1"
            label="Bar"
            value="bar"
          />
        </Picker>
        <Button
          onPress={[Function]}
          title="Done"
        />
      </View>
    </View>
  </Component>
</Fragment>
`;
